### PR TITLE
2024-08-27 update : prep for inf. wout dist.

### PR DESCRIPTION
**Pull request by cbernalz**
Regarding https://github.com/CDCgov/ww-inference-model/issues/96#issue-2487415321 .

### CODE REVIEW

File: inst/stan/wwinference.stan [hunk 1]
```diff
@@ -167,6 +168,7 @@ parameters {
   real log_scaling_factor;
   matrix[n_subpops-1,n_weeks] non_cent_spatial_dev_ns_mat;
   vector[n_weeks] norm_vec_aux_site;
+  cholesky_factor_corr[n_subpops-1] L_Omega;
```
Comment (dylanhmorris) [inline inst/stan/wwinference.stan:171]: This should be defined only if we're in the "inferred correlation but no distance matrix case". The way to do that in stan is to make it size zero in other cases.

See https://discourse.mc-stan.org/t/if-else-statement-inside-parameter-block/13937/2 for details

File: inst/stan/wwinference.stan [hunk 2]
```diff
@@ -344,6 +350,7 @@ model {
     log_sigma_generalized ~ normal(log_sigma_generalized_mu_prior, log_sigma_generalized_sd_prior);
     log_phi ~ normal(log_phi_mu_prior, log_phi_sd_prior);
     log_scaling_factor ~ normal(log_scaling_factor_mu_prior, log_scaling_factor_sd_prior);
+    L_Omega ~ lkj_corr_cholesky(2.0);
```
Comment (dylanhmorris) [inline inst/stan/wwinference.stan:353]: Related to the above, wrap this in an `if`

**Review by dylanhmorris — CHANGES_REQUESTED** [review]
A couple comments on the Stan side of things @cbernalz

File: R/get_stan_data.R [hunk 3]
```diff
@@ -379,6 +385,11 @@ get_stan_data <- function(input_count_data,
   #   not given.
   if (bool_spatial_comp) {
     ind_corr_func <- 0L
+    if (bool_spatial_corr_struct_exp) {
+      exp_corr_func <- 1
```
Comment (dylanhmorris) [inline R/get_stan_data.R:389]:
```suggestion
      exp_corr_func <- 1L
```

File: R/get_stan_data.R [hunk 4]
```diff
@@ -379,6 +385,11 @@ get_stan_data <- function(input_count_data,
...
+    } else {
+      exp_corr_func <- 0
```
Comment (dylanhmorris) [inline R/get_stan_data.R:391]:
```suggestion
      exp_corr_func <- 0L
```

File: R/get_stan_data.R [hunk 5]
```diff
@@ -387,6 +398,7 @@ get_stan_data <- function(input_count_data,
       nrow = subpop_data$n_subpops - 1,
       ncol = subpop_data$n_subpops - 1
     )
+    exp_corr_func <- 0
```
Comment (dylanhmorris) [inline R/get_stan_data.R:401]:
```suggestion
    exp_corr_func <- 0L
```

**Review by dylanhmorris — COMMENTED** [review]
Reflecting on this I think instead of two boolean flags (`ind_corr_func` and `exp_cor_func`) I'd just have a single integer `coff_func`, variable (`lower=0, upper=2>` where `0=ind, 1=lkj, 2=exp`, etc. This will also make it easier to extend in the future.

And then in the R function I'd have a string argument named "site_correlation_model" with choices `"iid"`,`lkj`, and `exponential`, or similar names (default `iid`)

**Review by kaitejohnson — COMMENTED** [review]
This looks really good to me, nice job implementing this, it looks really clean!

File: R/get_stan_data.R [hunk 6]
```diff
@@ -472,7 +484,8 @@ get_stan_data <- function(input_count_data,
     log_scaling_factor_mu_prior = params$log_scaling_factor_mu_prior,
     log_scaling_factor_sd_prior = params$log_scaling_factor_sd_prior,
     dist_matrix = dist_matrix,
-    ind_corr_func = ind_corr_func
+    ind_corr_func = ind_corr_func,
+    exp_corr_func = exp_corr_func
```
[the remark this answers is not in the document]
Comment (cbernalz) [reply R/get_stan_data.R:488]: @kaitejohnson I was just implementing what @dylanhmorris suggested.  I am having the string in `get_stan_data.R` where it is "iid" then ind_corr_func is set to 1 / exp_corr_func is set to 0 / ljk_corr_func is set to 0.  If the string is set to "exponential" then ind_corr_func is set to 0 / exp_corr_func is set to 1 / ljk_corr_func is set to 0.  If the string is set to "lkj" then ind_corr_func is set to 0 / exp_corr_func is set to 0 / ljk_corr_func is set to 1.  I just repushed @dylanhmorris's change request, can you take a look at it to see what we can do with this idea?

File: R/get_stan_data.R [hunk 7]
```diff
@@ -377,16 +376,27 @@ get_stan_data <- function(input_count_data,
...
+    if_ind_corr_func <- 0L
+    if_exp_corr_func <- 1L
+    if_lkj_corr_func <- 0L
+  } else if (corr_func == "lkj") {
+    if_ind_corr_func <- 0L
+    if_exp_corr_func <- 0L
+    if_lkj_corr_func <- 1L
+  } else {
+    stop("Correlation function desired currently not implemented\n
+         * must be either iid, exponential, lkj*")
```
Comment (dylanhmorris) [inline R/get_stan_data.R]:
```suggestion
         * must be one of 'iid', 'exponential', 'lkj'*")
```

File: inst/stan/wwinference.stan [hunk 8]
```diff
@@ -228,15 +231,22 @@ transformed parameters {
...
+  else if (if_exp_corr_func == 1){
     non_norm_omega = exponential_decay_corr_func(norm_dist_matrix, phi, l);
     norm_omega = matrix_normalization(non_norm_omega);
   }
+  else if (if_lkj_corr_func == 1) {
+    non_norm_omega = multiply_lower_tri_self_transpose(L_Omega);
+    norm_omega = matrix_normalization(non_norm_omega);
+  }
+  else {
+    print("Model will not get to there, get_stan_data.R provides protections");
```
Comment (dylanhmorris) [inline inst/stan/wwinference.stan:246]: Still worth raising an error even if something should not be occur.
```suggestion
    reject("Unsupported choice of correlation function. Check the data passed to Stan");
```

**Review by dylanhmorris — DISMISSED** [review]
LGTM

**Review by kaitejohnson — CHANGES_REQUESTED** [review]
Just requested changes to the language a bit. otherwise this looks great!

**Review by kaitejohnson — APPROVED** [review]
This looks good to me, thanks for making that language change I think it is more clear!

File: R/get_stan_data.R [hunk 9]
```diff
@@ -125,9 +125,10 @@ get_input_ww_data_for_stan <- function(preprocessed_ww_data,
 #' distance-based correlation function for epsilon. If NULL, use an independence
 #' correlation function, for current implementation (i.e. all sites' epsilon
 #' values are independent and identically distributed) .
-#' @param bool_spatial_comp Switch for whether or not infer
-#' site-to-site/"spatial" correlation matrix, currently correlation matrix
-#' follows exponential correlation structure.
+#' @param corr_structure_switch Integer variable to define the type of
+#' correlation matrix structure used.  Input 0 for an iid correlation structure,
+#' 1 for an exponential correlation structure based off distance matrix, and 3
```
Comment (dylanhmorris) [inline R/get_stan_data.R:130]:
```suggestion
#' 1 for an exponential correlation structure based off distance matrix, and 2
```

File: R/get_stan_data.R [hunk 10]
```diff
@@ -377,17 +378,18 @@ get_stan_data <- function(input_count_data,
...
     # This dist_matrix will not be used, only needed for stan data specs.
     dist_matrix <- matrix(
       0,
       nrow = subpop_data$n_subpops - 1,
       ncol = subpop_data$n_subpops - 1
     )
   }
+  if (!(corr_structure_switch %in% c(0, 1, 2))) {
+    stop("Correlation structure desired currently not implemented\n
+         * must be either 0-iid, 1-exponential, 2-lkj*")
```
Comment (dylanhmorris) [inline R/get_stan_data.R:391]:
```suggestion
    cli::cli_abort(c("Correlation structure requested currently not implemented.",
    Must be one of 0 (iid), 1 (exponential), and (2) lkj. Got {corr_structure_switch}"))
```

File: R/wwinference.R [hunk 11]
```diff
@@ -44,9 +44,10 @@
 #' distance-based correlation function for epsilon. If NULL, use an independence
 #' correlation function, for current implementation (i.e. all sites' epsilon
 #' values are independent and identically distributed) .
-#' @param bool_spatial_comp Switch for whether or not infer
-#' site-to-site/"spatial" correlation matrix, currently correlation matrix
-#' follows exponential correlation structure.
+#' @param corr_structure_switch Integer variable to define the type of
+#' correlation matrix structure used.  Input 0 for an iid correlation structure,
+#' 1 for an exponential correlation structure based off distance matrix, and 3
```
Comment (dylanhmorris) [inline R/wwinference.R:49]:
```suggestion
#' 1 for an exponential correlation structure based off distance matrix, and 2
```

File: inst/stan/wwinference.stan [hunk 12]
```diff
@@ -228,15 +229,22 @@ transformed parameters {
...
+  else if (corr_structure_switch == 1){
     non_norm_omega = exponential_decay_corr_func(norm_dist_matrix, phi, l);
     norm_omega = matrix_normalization(non_norm_omega);
   }
+  else if (corr_structure_switch == 2) {
+    non_norm_omega = multiply_lower_tri_self_transpose(L_Omega);
+    norm_omega = matrix_normalization(non_norm_omega);
+  }
+  else {
+    print("Model will not get to there, get_stan_data.R provides protections");
```
Comment (dylanhmorris) [inline inst/stan/wwinference.stan:246]:
```suggestion
    reject("Model should not reach this point. Invalid corr_structure_switch value. Check model code");
```

**Review by dylanhmorris — CHANGES_REQUESTED** [review]
Very close, thanks @cbernalz! Just a few small typo fixes and edits to make error messages clearer. Let me know when ready for re-review

**Review by dylanhmorris — APPROVED** [review]
LGTM

**Pull request merged**